js_modules/dagster-ui/packages/ui-core/src/runs/useCursorAccumulatedQuery.tsx

```diff
@@ -0,0 +1,45 @@
+import {useQuery} from '@apollo/client';
+import {DocumentNode} from 'graphql';
+import {useEffect, useState} from 'react';
+
+import {useQueryRefreshAtInterval} from '../app/QueryRefresh';
+
+type FetchState = {
+  hasMore: boolean;
+  cursor: string | null | undefined;
+};
+
+export function useCursorAccumulatedQuery<T, TVars extends {cursor?: string | null}, U>(options: {
```

Comment:
I'm always wary of using react state / effects as control flow for something complex like this. Itd be nicer to have useCursorAccumulatedQuery live outside of React. Maybe we could convert https://github.com/dagster-io/dagster/blob/7d67f72e405a56988807ba049e1c4eb0365be5d4/js_modules/dagster-ui/packages/ui-core/src/runs/fetchPaginatedBucketData.ts#L55 to a generator that yields dataSoFar after each cursor fetch so that you can render the partial data? also the way we're using useQueryRefreshAtInterval here means the logic that throttles the refresh interval based on how long it takes to fetch doesn't end up kicking in. Not blocking feedback though

Reply:
Hmm, the current version actually does render the partial data since it appends the results via `setFetched` as it fetches each page - `fetched` gets larger and larger as it pulls down pages, but the loading state disappears as soon as the first page is loaded.

The thing about the fetchPaginatedData version is that it doesn't stop if you click away from the tab...

I'll make a follow-up this afternoon and try to merge these together -- the ideal solution is probably a generator-version of `fetchPaginatedData` that has a callback for handling the partially fetched data and returns an handle to a `stop()` method.

Reply:
Hmm went ahead and tried this, and I think the challenge is the transition to the idle behavior with a `refreshState` like we usually have. I think we want the paginating fetch to resume if multiple pages of logs come in at once, and then fall back to the every-10s fetch once hasMore is false again, but have the `refreshState` linked up so you can click "Refresh" to trigger it / see the "Loading..."  state, etc.

I think we can make the outside-react fetcher vend a refreshState via an event emitter sort of pattern, and then store that in React state.

This is what I've got so far, but without the refreshState, going to merge this and break this into a follow-up:

```

export function buildPaginatingDataFetcher<DataType, CursorType, ErrorType>(
  fetchData: (cursor: CursorType | undefined) => Promise<{
    data: DataType[];
    hasMore: boolean;
    cursor: CursorType | undefined;
    error: ErrorType;
  }>,
) {
  let hasMoreData = true;
  let dataSoFar: DataType[] = [];
  let pollForData: NodeJS.Timeout;
  let currentCursor: CursorType | undefined = undefined;
  let stopped: boolean = false;

  return {
    start: (onDataSoFarChanged: (data: DataType[]) => void) => {
      if (stopped) {
        throw new Error('PaginatingDataFetcher cannot be restarted');
      }
      const _start = async () => {
        while (hasMoreData) {
          const {cursor, hasMore, data, error} = await fetchData(currentCursor);
          if (error) {
            throw error;
          }
          if (stopped) {
            break;
          }
          currentCursor = cursor;
          hasMoreData = hasMore;
          if (data.length > 0) {
            dataSoFar = dataSoFar.concat(data);
            onDataSoFarChanged(dataSoFar);
          }
        }
        pollForData = setTimeout(() => {
          hasMoreData = true;
          _start();
        }, 15 * 1000);
      };
      void _start();
    },
    stop: () => {
      clearTimeout(pollForData);
      stopped = true;
    },
  };
}

export function useCursorAccumulatedQuery<T, TVars extends {cursor?: string | null}, U>({
  query,
  variables,
  getResultArray,
  getNextFetchState,
}: {
  query: DocumentNode;
  variables: Omit<TVars, 'cursor'>;
  getResultArray: (result: T | undefined) => U[];
  getNextFetchState: (result: T | undefined) => FetchState | null;
}) {
  const [fetched, setFetched] = useState<U[] | null>(null);
  const client = useApolloClient();

  useEffect(() => {
    const fetcher = buildPaginatingDataFetcher(async (currentCursor: string | undefined) => {
      const resp = await client.query<T, TVars>({
        variables: {...variables, cursor: currentCursor},
        query,
      });

      return {...getNextFetchState(resp.data), data: getResultArray(resp.data), error: null};
    });

    fetcher.start(setFetched);

    return () => {
      fetcher.stop();
    };
  }, [query, client, variables, getResultArray, getNextFetchState]);

  return {fetched, refreshState};
}

```